File: Intersect.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

//Time Complexity: O(n).
//Space Complexity: O(1).
//This problem can be solved using different approaches, this approach reduces the extra space required. 

public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA == null || headB == null) return null;
        
        ListNode curr = headA;
        int lenA = 0;
        while(curr != null){
            curr = curr.next;
            lenA ++;
        }
        
        curr = headB;
        int lenB = 0;
        while(curr != null){
            curr = curr.next;
            lenB ++;
        }
        
        while(lenA > lenB){
            headA = headA.next;
            lenA --;
        }
        
        System.out.println(headA.val);
        
        while(lenB > lenA){
            headB = headB.next;
            lenB--;
        }
        
        while(headA != headB){
            headA = headA.next;
            headB = headB.next;
        }
        
        return headA;
    }
}

File: intersect.py
#Time complexity: O(n)
#Space complexity: O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        curr = headA
        lenA = 0
        while curr is not None:
            curr = curr.next
            lenA += 1
        curr = headB
        lenB = 0
        while curr is not None:
            curr = curr.next
            lenB += 1
        while lenA > lenB:
            headA = headA.next
            lenA -= 1
        while lenB > lenA:
            headB = headB.next
            lenB -= 1
        while headA != headB:
            headA = headA.next
            headB = headB.next
        
        return headA
